src/snappy_device_agents/devices/multi/tfclient.py

```diff
@@ -0,0 +1,129 @@
+import logging
+import urllib.parse
+import requests
+
+logger = logging.getLogger()
+
+
+class TFClient:
+    """Testflinger connection class"""
+
+    def __init__(self, url):
+        """Initialize the client with the url of the server
+
+        :param url: URL of the Testflinger server
+        """
+        self.server = url
+
+    def get(self, uri_frag, timeout=15):
+        """Submit a GET request to the server
+        :param uri_frag:
+            endpoint for the GET request
+        :return:
+            String containing the response from the server
+        """
+        uri = urllib.parse.urljoin(self.server, uri_frag)
+        try:
+            req = requests.get(uri, timeout=timeout)
+        except requests.exceptions.ConnectionError:
+            logger.error("Unable to communicate with specified server.")
+            raise
+        except IOError:
+            # This should catch all other timeout cases
+            logger.error(
+                "Timeout while trying to communicate with the server."
+            )
+            raise
+
+        try:
+            # If anything else went wrong, raise the proper exception
+            req.raise_for_status()
+        except OSError:
+            logger.error(
+                "Received status code %s from server.", req.status_code
+            )
+            raise
+        return req.text
+
+    def post(self, uri_frag, data, timeout=15):
+        """Submit a POST request to the server
+        :param uri_frag:
+            endpoint for the POST request
+        :return:
+            String containing the response from the server
+        """
+        uri = urllib.parse.urljoin(self.server, uri_frag)
+        try:
+            req = requests.post(uri, json=data, timeout=timeout)
+        except requests.exceptions.ConnectTimeout:
+            logger.error("Timout while trying to communicate with the server.")
+            raise
+        except requests.exceptions.ConnectionError:
+            logger.error("Unable to communicate with specified server.")
+            raise
+
+        try:
+            # If anything else went wrong, raise the proper exception
+            req.raise_for_status()
+        except OSError:
+            logger.error(
+                "Received status code %s from server.", req.status_code
+            )
+            raise
+        return req.text
+
+    def get_status(self, job_id):
+        """Get the status of a test job
+
+        :param job_id:
+            ID for the test job
+        :return:
+            String containing the job_state for the specified ID
+            (waiting, setup, provision, test, reserved, released,
+            cancelled, complete)
+        """
+        try:
+            endpoint = f"/v1/result/{job_id}"
+            data = json.loads(self.get(endpoint))
+            state = data.get("job_state")
+        except OSError:
+            logger.exception("Unable to get status for job %s", job_id)
+            state = "unknown"
```

Comment:
Shouldn't we also raise the exception here, as you do it in the `get` and `post` methods?

Reply:
We don't consider getting the state to be something that we should risk crashing on. If it fails, for any reason, then we handle it by returning that the state of the job is "unknown", rather than propagating an exception.